src/Raven.Client/Http/FailoverNodeSelector.cs

```diff
@@ -0,0 +1,88 @@
+using System;
+using System.Threading;
+
+namespace Raven.Client.Http
+{
+    public class FailoverNodeSelector : INodeSelector
+    {
+        protected Topology _topology;
+
+        public Topology Topology => _topology;
+
+        protected int _currentNodeIndex;
+
+        public FailoverNodeSelector(Topology topology)
+        {
+            _topology = topology;
+        }
+
+        public int GetCurrentNodeIndex()
+        {
+            return _currentNodeIndex;
+        }
+
+        public virtual void OnSucceededRequest(int nodeIndex)
+        {
+            //in failover mode nothing to do
+        }
+
+        public virtual void OnFailedRequest(int nodeIndex)
+        {
+            if (Topology.Nodes.Count == 0)
+                RequestExecutor.ThrowEmptyTopology();
+
+            var nextNodeIndex = nodeIndex < Topology.Nodes.Count - 1 ? nodeIndex + 1 : 0;
+            Interlocked.CompareExchange(ref _currentNodeIndex, nextNodeIndex, nodeIndex);
```

Comment:
If two threads hits this in the same time? You are going to get an invalid nextNodeIndex. Need to use the return value of the `CompareExchange`